### PR TITLE
Added more MC analysis code

### PR DESCRIPTION
**Pull request by espottesmith**
*Added functionality to analyze Monte Carlo runs in mrnet.stochastic:*
- generate_time_dep_profiles() - creates snapshots of the system state throughout simulations, how many times each reaction has fired, and the final state of the system
- final_state_analysis() - Generates statistics on the final states generated by generate_time_dep_profiles
- rank_reaction_counts() - Generates similar statistics for reactions

Unrelated to these changes, I added a flag for SerializedReactionNetwork. If use_thermo_cost is True (default), then a thermodynamics-based rate will be used. If it is False, then the k_A and k_B from the Reaction object will be used instead. This allows real kinetic information to be used in MC simulations.

### CODE REVIEW

File: src/mrnet/network/reaction_generation.py [hunk 1]
```diff
@@ -118,9 +118,19 @@ def __init__(
         self,
         input_entries,
         single_elem_interm_ignore=["C1", "H1", "O1", "Li1", "P1", "F1"],
+        electron_free_energy=-2.15,
+        temperature=298.15,
```
Comment (samblau) [inline src/mrnet/network/reaction_generation.py:122]: Should this point to ROOM_TEMP now? Also do we have a standardized default for electron free energy? if not, should we define one like we have for room temp?

File: src/mrnet/stochastic/serialize.py [hunk 2]
```diff
@@ -129,43 +130,49 @@ def __extract_index_mappings(self, reactions):
...
                 {
                     "reactants": product_indices,
                     "products": reactant_indices,
                     "free_energy": backward_free_energy,
+                    "rate_constant": backward_rate,
                 }
             )
 
-        for reaction in index_to_reaction:
+        if use_thermo_cost:
```
Comment (samblau) [inline src/mrnet/stochastic/serialize.py:153]: Pretty soon we're probably going to want to be able to use true barriers for some reactions and "thermo_cost" barriers for others, so this is going to need to change, but it's probably fine for now

Comment (espottesmith) [reply src/mrnet/stochastic/serialize.py:153]: I don't think there's a clean way to selectively use thermo/constant barrier costs within the serialization script. Do we want the user to have to provide an array of Booleans saying which reactions should and shouldn't have thermo costs?

I think the solution here is just to create a new RateCalculator class that provides a constant barrier, and apply that to Reactions without known kinetics.

File: src/mrnet/network/reaction_generation.py (hunk 1 above)
Comment (espottesmith) [reply src/mrnet/network/reaction_generation.py:122]: Yep. I thought I had changed everything before... maybe this came from a merge with Daniel's work? In any case, will fix.

Sent you a private message on electron_free_energy. I think default values make sense, but the defaults that make sense in general don't make sense for our work, and vice versa. This is one of those "are we building this for us, or for the community?" type questions.

**Pull request merged**